Improving OnUnknownFileFormat behavior in DatabaseChangeLog#include

momosetkn opened the pull request:
## Impact

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
<!--  Maintainers only: Mandatory Labels to add to a PR

At least one of the labels must be added to the PR before it's merged. If no label is provided the workflow will fail and you will not be able to merge the PR. After the label is added it re-runs the `Pull Request Labels / label (pull_request)` and gives a green check. 

`skipReleaseNotes`   - Don't show up on the Draft Release Notes page
`notableChanges`     - Any notable changes
`TypeEnhancement`    - New features
`TypeTest`           - New Test features
`TypeBug`            - bug fixes
`breakingChanges`    - any breaking changes
`APIBreakingChanges` - any API breaking changes
`sdou`               - Security, Driver and Other Updates -dependabot PR's
`newContributors`    - New Contributors 

-->


## Description

<!--
A clear and concise description of the change being made.  

- Introduce what was/will be done in the title
  - Titles show in release notes and search results, so make them useful
  - Use verbs at the beginning of the title, such as "fix", "implement", "improve", "update", and "add" 
  - Be specific about what was fixed or changed
  - Good Example: `Fix the --should-snapshot-data CLI parameter to be preserved when the --data-output-directory property is not specified in the command.`
  - Bad Example: `Fixed --should-snapshot-data`  
- If there is an existing issue this addresses, include "Fixes #XXXX" to auto-link the issue to this PR
- If there is NOT an existing issue, consider creating one.
  - In general, issues describe wanted change from an end-user perspective and PRs describe the technical change.
  - If this change is very small and not worth splitting off an issue, include `Steps To Reproduce`, `Expected Behavior`, and `Actual Behavior` sections in this PR as you would have in the issue.
- Describe what users need and how the fix will affect them
- Describe how the code change addresses the problem
- Ensure private information is redacted.
-->

- ignore a hidden file or version control system file.
hidden file or version control system file are shouldBe not target on `include` function.
Made a fix to exclude hidden files or version control system file from processing.
- decision-logic use should be filename without path.
- Additionally, modified to not logging the `OnUnknownFileFormat` for hidden files or version control system file.
- add test-code for onUnknownFileFormat on `include` method.
- improving OnUnknownFileFormat.WARN log message.

## Things to be aware of

<!--
- Describe the technical choices you made
- Describe impacts on the codebase
-->



## Things to worry about

<!--
- List any questions or concerns you have with the change
- List unknowns you have 
-->

## Additional Context

<!--
Add any other context about the problem here.
-->

i want ignore hoge/hoge2/.gitkeep in `include` method.

momosetkn committed:
Improving DatabaseChangeLogTest.groovy

momosetkn committed:
Improving DatabaseChangeLogTest.groovy

momosetkn committed:
Improving DatabaseChangeLogTest.groovy

momosetkn committed:
Add test-case to DatabaseChangeLogTest.groovy

momosetkn committed:
Add test-case to DatabaseChangeLogTest.groovy

momosetkn committed:
Add test-case to DatabaseChangeLogTest.groovy

momosetkn committed:
Add test-case to DatabaseChangeLogTest.groovy

momosetkn committed:
Improving DatabaseChangeLog

momosetkn committed:
Improving DatabaseChangeLog

momosetkn committed:
Improving DatabaseChangeLog

momosetkn commented on liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java:
```diff
@@ -973,7 +974,9 @@ public boolean include(String fileName,
                            ModifyChangeSets modifyChangeSets)
             throws LiquibaseException {
-        if (".svn".equalsIgnoreCase(fileName) || "cvs".equalsIgnoreCase(fileName)) {
+        String filenameWithoutPath = Paths.get(fileName).getFileName().toString();
+        boolean matchesHiddenFilename = HIDDEN_FILENAME_PATTERN.matcher(filenameWithoutPath).matches();
+        if ("cvs".equalsIgnoreCase(filenameWithoutPath) || matchesHiddenFilename) {
```
- ignore a hidden file or version control system file.
- decision-logic use should be filename without path.

momosetkn commented on liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java:
```diff
@@ -1031,11 +1034,9 @@ public boolean include(String fileName,
                 Scope.getCurrentScope().getLog(getClass()).warning(
-                        "included file " + normalizedFilePath + "/" + normalizedFilePath + " is not a recognized file type"
+                        "included file " + normalizedFilePath + " is not a recognized file type"
                 );
```
not logging the OnUnknownFileFormat for hidden files or version control system file.

momosetkn commented on liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java:
```diff
@@ -1031,11 +1034,9 @@ public boolean include(String fileName,
+            if (onUnknownFileFormat == OnUnknownFileFormat.WARN) {
                 Scope.getCurrentScope().getLog(getClass()).warning(
-                        "included file " + normalizedFilePath + "/" + normalizedFilePath + " is not a recognized file type"
+                        "included file " + normalizedFilePath + " is not a recognized file type"
```
improving OnUnknownFileFormat.WARN log message.
normalizedFilePath variable was duplicated.

momosetkn commented on liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy:
- add test-code for onUnknownFileFormat on include method.

momosetkn committed:
Merge branch 'master' into feat/improving_DatabaseChangeLog_include

momosetkn committed:
Merge branch 'master' into feat/improving_DatabaseChangeLog_include

momosetkn committed:
Update DatabaseChangeLogTest

momosetkn committed:
Update DatabaseChangeLogTest

momosetkn committed:
Update DatabaseChangeLogTest

momosetkn committed:
Update DatabaseChangeLogTest

MalloD12 committed:
Merge branch 'master' into feat/improving_DatabaseChangeLog_include

MalloD12 committed:
File name path normalization added.

MalloD12 committed:
File name path normalization added.

MalloD12 committed:
File name path normalization added.

MalloD12 reviewed (dismissed):
Code changes were overall looking good to me. Although, I have to make the below code change because it seems path were not normalized and some tests were failing on Windows:

```java
if(fileName.startsWith("classpath:") && !GlobalConfiguration.PRESERVE_CLASSPATH_PREFIX_IN_NORMALIZED_PATHS.getCurrentValue()) {
    filenameWithoutPath = Paths.get(normalizePath(fileName)).getFileName().toString();
} else {
    filenameWithoutPath = fileName;
}
```

MalloD12 commented:
> @MalloD12 What is the reason for including classpath conditions like bellow code?
> 
> ```java
> if(fileName.startsWith("classpath:") && !GlobalConfiguration.PRESERVE_CLASSPATH_PREFIX_IN_NORMALIZED_PATHS.getCurrentValue()) {
> ```

Hi @momosetkn,

There were some tests (UrlDrivenResourceAccessorTest) that were failing on Windows because the path was not being normalized (classpath prefix not removed, which was making changelos to be not found). I added the `normalizePath()` method. Still, then there was another method failing because, in that particular case `PRESERVE_CLASSPATH_PREFIX_IN_NORMALIZED_PATHS` was being set as `true` and because of the normalize call I added that the test was failing, and that's the why of the IF logic I added. If you are happy with these code changes, please let's discuss it here, or feel free to change as needed and then we will review it with the team.

Thanks for your contribution with this PR,
Daniel.

momosetkn committed:
Log assertion added.

momosetkn committed:
Log assertion added.

momosetkn committed:
Log assertion added.

momosetkn committed:
Log assertion added.

momosetkn commented:
@MalloD12 I added WARN log assertion in test. Should be not WARN logging when `cvs` and `.*` files. but the test failed.

MalloD12 commented:
> @MalloD12 I added WARN log assertion in test. Should be not WARN logging when `cvs` and `.*` files. but the test failed.

Hi @momosetkn,

Were these tests failing? I'm pretty sure I saw all tests passing only ones I saw failing were the ones I mentioned to you and the ones where we realized we needed to do some additional changes. I see now tests are failing with the below error message:

```java
Error:    DatabaseChangeLogTest.include ignore if version control system files:465 Condition not satisfied:

bufferLog.getLogAsString(Level.WARNING) == ""
|         |              |     |        |
|         |              |     WARNING  false
|         |              |              36602 differences (0% similarity)
|         |              |              ([2025-03-24T13:19:59.399] WARNING included file com/example/cvs is not a recognized file type\nliquibase.exception.UnknownChangelogFormatException: Cannot find parser that supports com/example/cvs\n\tat liquibase.parser.ChangeLogParserFactory.getParser(ChangeLogParserFactory.java:59)\n\tat liquibase.changelog.DatabaseChangeLog.include(DatabaseChangeLog.java:1085)\n\tat org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)\n\tat liquibase.changelog.DatabaseChangeLogTest$2.run(DatabaseChangeLogTest.groovy:454)\n\tat liquibase.Scope.lambda$child$0(Scope.java:212)\n\tat liquibase.Scope.child(Scope.java:221)\n\tat liquibase.Scope.child(Scope.java:211)\n\tat liquibase.Scope.child(Scope.java:190)\n\tat org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)\n\tat liquibase.changelog.DatabaseChangeLogTest.$spock_feature_0_10(DatabaseChangeLogTest.groovy:451)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat org.spockframework.util.ReflectionUtil.invokeMethod(ReflectionUtil.java:187)\n\tat org.spockframework.runtime.model.MethodInfo.lambda$new$0(MethodInfo.java:49)\n\tat org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:156)\n\tat org.spockframework.runtime.PlatformSpecRunner.invokeRaw(PlatformSpecRunner.java:407)\n\tat org.spockframework.runtime.PlatformSpecRunner.invoke(PlatformSpecRunner.java:390)\n\tat org.spockframework.runtime.PlatformSpecRunner.runFeatureMethod(PlatformSpecRunner.java:324)\n\tat org.spockframework.runtime.IterationNode.execute(IterationNode.java:50)\n\tat org.spockframework.runtime.SimpleFeatureNode.execute(SimpleFeatureNode.java:58)\n\tat org.spockframework.runtime.SimpleFeatureNode.execute(SimpleFeatureNode.java:15)\n\tat org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:156)\n\tat org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)\n\tat org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:146)\n\tat org.spockframework.runtime.SpockNode.sneakyInvoke(SpockNode.java:40)\n\tat org.spockframework.runtime.IterationNode.lambda$around$0(IterationNode.java:67)\n\tat org.spockframework.runtime.PlatformSpecRunner.lambda$createMethodInfoForDoRunIteration$5(PlatformSpecRunner.java:236)\n\tat org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:156)\n\tat org.spockframework.runtime.PlatformSpecRunner.invokeRaw(PlatformSpecRunner.java:407)\n\tat org.spockframework.runtime.PlatformSpecRunner.invoke(PlatformSpecRunner.java:390)\n\tat org.spockframework.runtime.PlatformSpecRunner.runIteration(PlatformSpecRunner.java:218)\n\tat org.spockframework.runtime.IterationNode.around(IterationNode.java:67)\n\tat org.spockframework.runtime.SimpleFeatureNode.lambda$around$0(SimpleFeatureNode.java:52)\n\tat org.spockframework.runtime.SpockNode.sneakyInvoke(SpockNode.java:40)\n\tat org.spockframework.runtime.FeatureNode.lambda$around$0(FeatureNode.java:41)\n\tat 
``` 

See full execution details [here](https://github.com/liquibase/liquibase/actions/runs/14024706611/job/39294907872?pr=6190).

Could you please take a look at this?

Thanks,
Daniel.

momosetkn commented:
@MalloD12 
Sorry, I don't understand what you mean. 

Ignore a hidden file(begining `.` start with file) or version control system file(`csv`, `.gitkeep`)  without puts WARN log. Because, these files are clearly unnecessary for Liquibase. Is there any problem with this?
If there is no problem, I fix the code to work on Windows, Linux, and macOS.

MalloD12 committed:
Merge remote-tracking branch 'origin/master' into momosetkn-feat/improving_DatabaseChangeLog_include

momosetkn commented:
@MalloD12 
First, let's talk about the specifications.
when hidden file(begining . start with file) or version control system file(csv, .gitkeep), put warn log?or not puts warn log? YES(puts warn log) or NO(not puts warn log).
My opinion is NO(not puts warn log).
The reason I added them test is because not need logging in these files. The original code(`cvs` and `.svn` file are ignored) was written like that.
https://github.com/liquibase/liquibase/blob/d81577309bc85f069f2d376f3329133ef8f0d8b9/liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy#L463-L465

`bufferLog.getLogAsString(Level.WARNING) == "" ` is verify that no output WARN log.

MalloD12 committed:
File name without path code refactoring.

MalloD12 committed:
File name without path code refactoring.

MalloD12 committed:
File name without path code refactoring.

MalloD12 commented:
> @MalloD12 First, let's talk about the specifications. when hidden file(begining . start with file) or version control system file(csv, .gitkeep), put warn log?or not puts warn log? YES(puts warn log) or NO(not puts warn log). My opinion is NO(not puts warn log). The reason I added them test is because not need logging in these files. The original code(`cvs` and `.svn` file are ignored) was written like that.
> 
> https://github.com/liquibase/liquibase/blob/d81577309bc85f069f2d376f3329133ef8f0d8b9/liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy#L463-L465
> 
> `bufferLog.getLogAsString(Level.WARNING) == "" ` is verify that no output WARN log.

Hey @momosetkn,

Sorry for the delay in my answer. I agree we should NOT log any warning message in the cases of cvs/hidden files, I made a little tweak to the code I added the other day, and seems to be ok now. Please review it and let me know if you have any comments/concerns about it.

Thanks,
Daniel.

filipelautert commented:
@MalloD12 is this pr ready for review?

MalloD12 commented:
@filipelautert, according to the previous discussions, it seems there is nothing pending. So, I think we can move forward with the review of this PR.  It seems I forgot to update the status of the review flow after making the last tweak from a discussion we had with @momosetkn. I'll change it now.

MalloD12 committed:
Merge branch 'master' into feat/improving_DatabaseChangeLog_include

filipelautert committed:
Merge branch 'master' into feat/improving_DatabaseChangeLog_include

momosetkn commented:
Sorry for the delay.

1. It seems that the coverage is not met for the added process.
![Screenshot from 2025-05-05 22-51-25](https://github.com/user-attachments/assets/0bc8ff01-70a4-4c31-9e0e-f1dd79913d1f)

2. I using deprecated `liquibase.logging.core.BufferedLogService` feature.  need to replace it.

MalloD12 commented:
Hi @momosetkn,

1. Will you be then adding some additional test/s for it?
2. It's right, `liquibase.logging.core.BufferedLogService` is a `deprecated` class, but for testing purposes, we are still using it. Although, if you want to make a replacement for a non-deprecated class that would be perfect too.
 
Thanks,
Daniel.

momosetkn commented:
@MalloD12 OK, I will add test-case.

momosetkn committed:
Fix hidden filename regex to match only leading dots

momosetkn committed:
Fix hidden filename regex to match only leading dots

momosetkn committed:
Add tests for handling hidden files and classpath in includes

momosetkn committed:
Add tests for handling hidden files and classpath in includes

momosetkn commented:
@MalloD12 I added test-case.

MalloD12 commented:
Awesome, @momosetkn! I'll look at it now.

Thanks,
Daniel.

MalloD12 commented on liquibase-standard/src/test/groovy/liquibase/changelog/DatabaseChangeLogTest.groovy:
```diff
@@ -432,6 +433,88 @@ create view sql_view as select * from sql_table;'''
+        bufferLog.getLogAsString(Level.WARNING) == ""
+    }
+
+    def "start with classpath: and cvs"() {
```
What are you trying to test with this scenario? Whether a file with classpath prefix and extension is a `csv`. Because I can see in the mock that we are relating this path( which is not a CSV file extension) to a `XML` file. I might not be getting the purpose of this scenario.

tati-qalified committed:
Merge branch 'master' into feat/improving_DatabaseChangeLog_include

tati-qalified reviewed (commented):
Hi @momosetkn, thank you for your contribution!
I tested this manually, and with `include` the logged output is `Cannot find parser that supports abc.cde`. Only `includeAll` shows `included file (...)`.
Do you know why this might be?

MalloD12 committed:
includeAll test added.

MalloD12 committed:
includeAll test added.

MalloD12 committed:
Merge branch 'feat/improving_DatabaseChangeLog_include' of https://github.com/momosetkn/liquibase into momosetkn-feat/improving_DatabaseChangeLog_include

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/liquibase/liquibase?pullRequest=6190) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/liquibase/liquibase?pullRequest=6190) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: filipelautert<br/>:white_check_mark: MalloD12<br/>:white_check_mark: tati-qalified<br/>:x: momosetkn<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/liquibase/liquibase?pullRequest=6190) it.</sub>

MalloD12 commented:
@tati-qalified can you please share your changelog/s you are using to perform your tests? That would be useful to debug and see what's still missing.

Thanks,
Daniel.

tati-qalified commented:
File structure:
```
changelog.xml
abc.def
abc.sql
changes 
      |_ script.sql
      |_ change.xml
      |_ git.ignore
      |_ txt.123
```

changelog.xml:
```
    <includeAll path="changes" relativeToChangelogFile="true" />
    <include file="abc.def" relativeToChangelogFile="true" />
```

The content of all other files isn't important, as the behaviour being tested should happen before the file's content is accessed.

MalloD12 committed:
- Update include call to avoid failing by default if changelog file extension is not recognized.
- Add OnUnknownChangeLogFileFormatTest to group all related OnUnknownFileFormat related tests.
- Remove OnUnknownFileFormat related tests from DatabaseChangeLogTest class.

MalloD12 committed:
Merge branch 'feat/improving_DatabaseChangeLog_include' of https://github.com/momosetkn/liquibase into momosetkn-feat/improving_DatabaseChangeLog_include

MalloD12 commented:
@momosetkn - @tati-qalified I have made the following changes to this PR after the last changes made by @momosetkn :

- The last reported issue was because we were invoking include with `OnUnknownFileFormat.FAIL`, which was why it was not showing any warning, and execution was getting stopped. So, to be consistent with `includeAll`, I changed it to `WARN`.
- Created a new test class to isolate `OnUnknownFileFormat` tests, and added new scenarios.
- Migreate `OnUnknownFileFormat` related tests from `DatabaseChangeLogTest` to `OnUnknownChangeLogFileFormatTest`.

@momosetkn, please review the changes made, and let me know if you have any questions/concerns with what has been done.

@tati-qalified, feel free to test again and let us know if you see any more cases not working as expected.

Thanks,
Daniel.

MalloD12 reviewed (approved):
This is ok from my side, it's ready for a second review and for testing.

tati-qalified commented on liquibase-standard/src/main/java/liquibase/changelog/DatabaseChangeLog.java:
```diff
@@ -657,7 +658,7 @@ private void handleInclude(ParsedNode node, ResourceAccessor resourceAccessor, M
                     ignore,
                     node.getChildValue(null, LOGICAL_FILE_PATH, String.class),
-                    OnUnknownFileFormat.FAIL,
+                    OnUnknownFileFormat.WARN,
```
should we even be making this change?

MalloD12 replied:
@mariochampion do you think we should revert this change? Current behavior is to make execution fail if we try to include a changelog with an invalid extension, while with this PR we are making it just `WARN` a message. `includeAll` will behave differently. I think the options are:

1. Revert this change (keep failing execution) if an invalid extension is used, but I think we should be changing the includeAll behavior to act in the same way. 
2. Progress with this PR as it is: both `include`/`includeAll` will warn if an invalid extension is found.

I think in any case, we should be documenting the implemented behavior, so this is clear and visible for users.

What do you think?

Thanks,
Daniel.
